include/jemalloc/internal/guard.h

```diff
@@ -0,0 +1,76 @@
+#ifndef JEMALLOC_INTERNAL_GUARD_H
+#define JEMALLOC_INTERNAL_GUARD_H
+
+#include "jemalloc/internal/ehooks.h"
+#include "jemalloc/internal/emap.h"
+
+#define PAGE_GUARDS_SIZE (2 * PAGE)
+
+#define SAN_GUARD_LARGE_N_EXTENTS 0
+#define SAN_GUARD_SMALL_N_EXTENTS 0
+
+extern size_t opt_san_guard_large_nextents;
+extern size_t opt_san_guard_small_nextents;
+
+void guard_pages(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata, emap_t *emap);
+void unguard_pages(tsdn_t *tsdn, ehooks_t *ehooks, edata_t *edata, emap_t *emap);
+void tsd_san_init(tsd_t *tsd);
+
+static inline bool
+san_enabled(void) {
+	return (opt_san_guard_large_nextents != 0 ||
+	    opt_san_guard_small_nextents != 0);
+}
+
+static inline bool
+large_extent_decide_guard(tsdn_t *tsdn, ehooks_t *ehooks, size_t size,
+    size_t alignment) {
+	if (opt_san_guard_large_nextents == 0 ||
+	    ehooks_guard_will_fail(ehooks) || tsdn_null(tsdn)) {
+		return false;
+	}
+
+	tsd_t *tsd = tsdn_tsd(tsdn);
+	uint64_t n = tsd_san_guard_large_nextent_get(tsd);
+	assert(n >= 1);
+	if (n > 1) {
+		/*
+		 * Subtract conditionally because the guard may not happen due
+		 * to alignment or size restriction below.
+		 */
+		*tsd_san_guard_large_nextentp_get(tsd) = n - 1;
```

Comment:
This is subtracted regardless of the success of the following allocation. What if the allocation fails? I don't know how relevant this scenario is, but if overcommit is turned off and the system is out of memory, maybe reducing the rate of guarded allocations would help?

In other words, have you considered the policy "guard every Nth **successful** allocation" vs "guard every Nth allocation attempt"?

Reply:
Agree, thanks. I was just curious what you thought about this.